pkg/rule/criteria.go

```diff
@@ -1,94 +1,286 @@
+	case UserID:
+		return "UserID"
+	case UserSubscription:
+		return "UserSubscription"
+	default:
+		return "unknown"
+	}
 }
 
-// MatcherBool defines methods for matching a rule on bool type
-type MatcherBool struct {
-	value bool
+// Criteria is a collection of Criterion.
+type Criteria []Criterion
+
+// Criterion is a single decision which can be evaluated to decide if a Rule
+// should be applied.
+type Criterion struct {
+	Comparator Comparator
+	Key        CriterionKey
+	Value      interface{}
+	Path       string
 }
 
-func (m MatcherBool) match(input interface{}) (bool, error) {
-	t, ok := input.(bool)
-	if !ok {
-		return false, errors.Wrapf(errors.ErrInvalidTypeToMatch, "missmatch %s != bool", reflect.TypeOf(input).Kind())
+// MarshalJSON to satisfy json.Marshaler.
+func (c Criterion) MarshalJSON() ([]byte, error) {
+	var value interface{}
+
+	switch c.Key {
+	case DeviceLocationLocale:
+		t, ok := c.Value.(language.Tag)
+		if !ok {
+			return nil, errors.Wrapf(errors.ErrInvalidTypeToMatch, "%s: value not a language.Tag", c.Key)
+		}
+
+		value = t.String()
+	case UserSubscription:
+		t, ok := c.Value.(int)
+		if !ok {
+			return nil, errors.Wrapf(errors.ErrInvalidTypeToMatch, "%s: value noat an int", c.Key)
+		}
+
+		value = t
+	case UserID:
+		t, ok := c.Value.([]string)
+		if !ok {
+			return nil, errors.Wrapf(errors.ErrInvalidTypeToMatch, "%s: value not a string slice", c.Key)
+		}
+
+		value = t
+	default:
+		return nil, errors.Errorf("marshaling for '%s' not supported", c.Key)
 	}
 
-	return t == m.value, nil
+	return json.Marshal(struct {
+		Comparator Comparator   `json:"comparator"`
+		Key        CriterionKey `json:"key"`
+		Value      interface{}  `json:"value"`
+		Path       string       `json:"path"`
+	}{
+		Comparator: c.Comparator,
+		Key:        c.Key,
+		Value:      value,
+		Path:       c.Path,
+	})
 }
 
-// MatcherInt defines methods for matching a rule on int type
-type MatcherInt struct {
-	Comparator comparator
-	Value      int
-}
+// UnmarshalJSON to satisfy json.Unmarshaler.
+func (c *Criterion) UnmarshalJSON(raw []byte) error {
+	v := struct {
+		Comparator Comparator   `json:"comparator"`
+		Key        CriterionKey `json:"key"`
+		Value      interface{}  `json:"value"`
+		Path       string       `json:"path"`
+	}{}
 
-func (m MatcherInt) match(input interface{}) (bool, error) {
-	t, ok := input.(int)
-	if !ok {
-		return false, errors.Wrapf(errors.ErrInvalidTypeToMatch, "missmatch %s != int", reflect.TypeOf(input).Kind())
+	if err := json.Unmarshal(raw, &v); err != nil {
+		return err
 	}
 
-	switch m.Comparator {
-	case comparatorGT:
-		return t > m.Value, nil
+	c.Comparator = v.Comparator
+	c.Key = v.Key
+	c.Path = v.Path
+
+	switch c.Key {
+	case DeviceLocationLocale:
+		s, ok := v.Value.(string)
+		if !ok {
+			return errors.Wrapf(errors.ErrInvalidTypeToMatch, "%s: value not a string", c.Key)
+		}
+
+		t, err := language.Parse(s)
+		if err != nil {
+			// TODO(xla): Wrap in proper marshal error.
```

Comment:
Would be nice to resolve this `TODO` and introduce a concrete error for this case.